### ISSUE TITLE
Why not develop this on Kitware's gitlab instance?

### ISSUE THREAD

**Issue by cristianadam** [stryku/cmakesl#225]
Is there any reason for not trying to upstream this?

There are discussions about having `lua` as a scripting language here: https://gitlab.kitware.com/cmake/cmake/issues/19863

cmakesl looks way more familiar to c/c++ users.

**Comment by stryku** [comment]
First of, thanks for pointing the Lua issue out.

Well, that's quite a long story. I'll try to summarize this up.

1. When I started thinking about implementing CMakeSL I knew that I can't just go to the CMake's gitlab and create issue like "I'd like to create a new language for CMake. It'll be like C++ and it'll be great". I wanted to start discussion with CMake developers when I have a good POC.
2. Like five months ago I 'released' a POC that can build a simple hello world application. I posted it on r/cpp, r/cmake and on CMake mailing list. There were discussions on reddit, but no response on the mailing list, so I thought that CMake devs are still to sceptical. I decided to continue project and implement it to the state where it can prove its usefulness. I decided to do this in two phases:
2.1 Implement it to the state that it's able to build abseil-cpp project.
2.2 Implement it to the state that it's able to build LLVM + Clang. After this, it'd be clear whether CMakeSL helps with maintaining big project with a lot of CMake code. I wanted to start a discussion after this.
3. Recently CMakeSL became advanced enough to build itself. Still not able to build abseil-cpp, though.
4. You created the issue.

To answer your question. Why it's not at the CMake gitlab? Simply because I don't know whether CMake developers think that CMakeSL is a thing and they want in the project. Additionally, I'm not sure whether we (me and everyone else) have enough knowledge to predict whether CMakeSL would help with writing CMakeLists etc. I mean I believie it would. But you know, I'm just a simple C++ dev.

Nevetheless, we'll find out. I'm going to add a comment about CMakeSL to the Lua issue, clean up CMakeSL repo (README, docs, examples etc.) and create an issue to discuss merging CMakeSL into CMake's codebase.

**Comment by cristianadam** [comment]
Thank you. I recently been given developer rights on CMake, and managed to get PCH and Unity builds in CMake 3.16, so I can say that one can get things into CMake.

You need to document, add tests, integrate into existing machinery. 

If cmakesl can somehow reuse clang-format and clang-tidy it would be very, very nice. CMake has a great backwards compatibility story, but it doesn't have tools to migrate old code to new code. CMakesl could have this with a tool like clang-tidy.

On the other hand people want to have something declarative, but this won't happen by itself. Somebody needs to do the work. We all agree that CMake language scripting is a pain, all those uppercase strings which need to pass semantic value, and you always have to look into documentation to get things right. 

If you have a way to have both scripting languages coexist, and a way to migrate from one to another, I don't think people will be against it. CMake people are pragmatic, and open to changes, you just have to be persistent enough to pass all the reviewing process.

**Comment by stryku** [comment]
Thanks for the tips.

clang-format already works. All *.cmsl files in this repo are formatted with it (with a style got from the .clang-format file).
Also, doxygen works with *.cmsl files under doc/ folder.
I haven't tried clang-tidy, though.

About declarative language. I too think that it's a good approach for simple stuff. But, I've talked with a couple of people and they say, that declarative language doesn't work so well in a really big project, when you need to do complicated stuff. You'd need to mix it with an imperative language at some point.

About mixing old CMake and CMakeSL, that's exactly what happens in this repo. As I said in the previous post, CMakeSL is able to build itself. googletest is used as a testing framework. It's cloned and its sources are directly in the repo, under external/googletest. Of course, its `CMakeLists` doesn't support CMakeSL, old `CMakeLists.txt` are present. In this repo, in root `CMakeLists.cmsl`, `add_subdirectory("external/googletest")` is called. So from CMakeSL script, you can add another subdirectory with and old `CMakeLists.txt`. [The call is here.](https://github.com/stryku/cmakesl/blob/master/CMakeLists.cmsl#L41)

It'll work the other way, which is, in an old `CMakeLists.txt`, adding subdirectory with a `CMakeList.cmsl`, but I've accidentally removed it (it was late, don't ask).

So basically you can migrate your project to CMakeSL in small steps, directory after directory.

**Comment by cristianadam** [comment]
That looks great. Another thing that CMakeSL can improve, is overloading. With CMake you can overload a macro and a function and call the `_` name, which is the old name. But you can do it only once. 

You cannot have `find_package` redefined in vcpkg, and then again in your own script.

Another point would be performance wise. No idea how CMakeSL performs. I guess I could give it a try at some point 😄

**Comment by stryku** [comment]
Thanks, there is a lot of stuff to introduce. We'll see whether there will be a chance to implement it (:

I've created the issue. It's here, if anyone is interested: https://gitlab.kitware.com/cmake/cmake/issues/19889
